packages/kbn-logging/src/appenders.ts

```diff
@@ -15,6 +15,7 @@ import { LogRecord } from './log_record';
  */
 export interface Appender {
   append(record: LogRecord): void;
+  update(config: { appenders?: Map<string, Appender> }): void;
```

Comment:
or appender might be passed from outside. I haven't checked out how it works, but the interface in java looks like it is  https://github.com/apache/log4j/blob/trunk/src/main/java/org/apache/log4j/rewrite/RewriteAppender.java#L74
```js
const appenderNames = appender.getAppenders?();
if(appenderNames) {
  appenderNames.forEach(name => appender.setAppender(this.appender.get(name))
}
```

Reply:
@restrry It looks like log4j has the concept of an [attachable appender](https://github.com/apache/log4j/blob/7be00eed88152dd011a619e8bae5a631235c3f4c/src/main/java/org/apache/log4j/spi/AppenderAttachable.java) internally, and they will take the appender refs from the configuration and [only attach those appenders to the rewrite appender](https://github.com/apache/log4j/blob/7be00eed88152dd011a619e8bae5a631235c3f4c/src/main/java/org/apache/log4j/xml/DOMConfigurator.java#L278-L284).

We could try to take a similar approach:
- Create `AttachableAppender` interface which extends from `Appender`, and have `RewriteAppender` implement that.
- `AttachableAppender` would need to have (at least) `addAppender`. I suppose it would also need a way to expose the required `appenderRefs` (otherwise we'd need to infer that from the computed config)
- Then instead of the lifecycle methods, in the logging system we could do something like:

```ts
    for (const [appenderKey, appenderConfig] of computedConfig.appenders) {
      this.appenders.set(appenderKey, Appenders.create(appenderConfig));
    }

    for (const [, appender] of this.appenders) {
      if (appender.isAttachable) {
        appender.appenderRefs.forEach(ref => appender.addAppender(this.appenders.get(ref));
      }
    }
```

WDYT? It would be nice to have the attached appenders scoped to only what is required and not need to mess with the `update` lifecycle, but OTOH it requires introducing the new concept of `AttachableAppender`

Reply:
>  It would be nice to have the attached appenders scoped to only what is required and not need to mess with the update lifecycle, but OTOH it requires introducing the new concept of AttachableAppender

What if we use duck-typing? Can get by with checking that `appender` has `addAppender` method?

Reply:
> What if we use duck-typing? Can get by with checking that appender has addAppender method?

Yep, duck-typing is an option too, though I think it'd still look similar to my example above, even if we don't have an official `AttachableAppender` interface.

The simplest approach would be to check if it has `addAppender`, and then add all appenders to it.

If we want to scope it to the required appender references, we could expose `appender.appenderRefs` or similar (as in example above), and use that list to only call `addAppender` with the required refs.

Reply:
I ended up duck-typing with `appenderRefs` and `addAppender`, though I did still add the methods as optional items in the appender interface. LMK what you think.